src/backend/fastapi_app/rag_simple.py

```diff
@@ -24,63 +103,127 @@ def __init__(
 
         return RetrievalResponse(
-            message=Message(content=str(first_choice_message.content), role=first_choice_message.role),
+            message=Message(
+                content=str(chat_completion_response.choices[0].message.content), role=AIChatRoles.ASSISTANT
+            ),
+            context=RAGContext(
+                data_points={item.id: item.to_dict() for item in results},
+                thoughts=[
+                    ThoughtStep(
+                        title="Search query for database",
+                        description=chat_params.original_user_query if chat_params.enable_text_search else None,
+                        props={
+                            "top": chat_params.top,
+                            "vector_search": chat_params.enable_vector_search,
+                            "text_search": chat_params.enable_text_search,
+                        },
+                    ),
+                    ThoughtStep(
+                        title="Search results",
+                        description=[result.to_dict() for result in results],
+                    ),
+                    ThoughtStep(
+                        title="Prompt to generate answer",
+                        description=[str(message) for message in contextual_messages],
+                        props=(
+                            {"model": self.chat_model, "deployment": self.chat_deployment}
+                            if self.chat_deployment
+                            else {"model": self.chat_model}
+                        ),
+                    ),
+                ],
+            ),
+        )
+
+    async def run_stream(
+        self,
+        messages: list[ChatCompletionMessageParam],
+        overrides: dict[str, Any] = {},
+    ) -> AsyncGenerator[RetrievalResponseDelta, None]:
+        chat_params = self.get_params(messages, overrides)
+
+        # Retrieve relevant items from the database
+        # Generate a contextual and content specific answer using the search results and chat history
+        contextual_messages, results = await self.retreive_and_build_context(chat_params=chat_params)
+
+        chat_completion_async_stream: AsyncStream[
+            ChatCompletionChunk
+        ] = await self.openai_chat_client.chat.completions.create(
+            # Azure OpenAI takes the deployment name as the model name
+            model=self.chat_deployment if self.chat_deployment else self.chat_model,
+            messages=contextual_messages,
+            temperature=chat_params.temperature,
+            max_tokens=chat_params.response_token_limit,
+            n=1,
+            stream=True,
+        )
+
+        # Forcefully close the database session before yielding the response
+        # Yielding keeps the connection open while streaming the response until the end
+        # The connection closes when it returns back to the context manger in the dependencies
+        await self.searcher.db_session.close()
```

Comment:
This needs your attention, I'm open to other suggestions. It currently works and fixes the SQLAlchemy warnings I get about open connection from the pool.

Reply:
```bash
  /usr/local/lib/python3.11/mimetypes.py:99: SAWarning: The garbage collector is trying to clean up non-checked-in connection <AdaptedConnection <asyncpg.connection.Connection object at 0xffff91569300>>, which will be terminated.  Please ensure that SQLAlchemy pooled connections are returned to the pool explicitly, either by calling ``close()`` or by using appropriate context managers to manage their lifecycle.
    exts = self.types_map_inv[strict].setdefault(type, [])

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```